core-plugins/src/test/java/co/cask/hydrator/plugin/batch/joiner/JoinerTestRun.java

```diff
@@ -0,0 +1,264 @@
+    // write input data
+    // 1: 1, matrix
+    // 2: 2, equilibrium
+    // 3: 3, avatar
+    DataSetManager<Table> filmManager = getDataset(filmDatasetName);
+    Table filmTable = filmManager.get();
+    Put put = new Put(Bytes.toBytes(1));
+    put.add("film_id", "1");
+    put.add("film_name", "matrix");
+    filmTable.put(put);
+
+    put = new Put(Bytes.toBytes(2));
+    put.add("film_id", "2");
+    put.add("film_name", "equilibrium");
+    filmTable.put(put);
+
+    put = new Put(Bytes.toBytes(3));
+    put.add("film_id", "3");
+    put.add("film_name", "avatar");
+    filmTable.put(put);
+    filmManager.flush();
+
+    // 1: 1, matrix, alex
+    // 2: 1, matrix, bob
+    // 3: 2, equilibrium, cathie
+    // 4: 3, avatar, samuel
+    DataSetManager<Table> filmActorManager = getDataset(filmActorDatasetName);
+    Table filmActorTable = filmActorManager.get();
+    put = new Put(Bytes.toBytes(1));
+    put.add("film_id", "1");
+    put.add("film_name", "matrix");
+    put.add("actor_name", "alex");
+    filmActorTable.put(put);
+
+    put = new Put(Bytes.toBytes(2));
+    put.add("film_id", "1");
+    put.add("film_name", "matrix");
+    put.add("actor_name", "bob");
+    filmActorTable.put(put);
+
+    put = new Put(Bytes.toBytes(3));
+    put.add("film_id", "2");
+    put.add("film_name", "equilibrium");
+    put.add("actor_name", "cathie");
+    filmActorTable.put(put);
+
+    put = new Put(Bytes.toBytes(4));
+    put.add("film_id", "3");
+    put.add("film_name", "avatar");
+    put.add("actor_name", "samuel");
+    filmActorTable.put(put);
+    filmActorManager.flush();
+
+    // 1: 1, matrix, action
+    // 2: 1, matrix, thriller
+    // 3: 2, equilibrium, action
+    DataSetManager<Table> filmCategoryManager = getDataset(filmCategoryDatasetName);
+    Table filmCategoryTable = filmCategoryManager.get();
+    put = new Put(Bytes.toBytes(1));
+    put.add("film_id", "1");
+    put.add("film_name", "matrix");
+    put.add("category_name", "action");
+    filmCategoryTable.put(put);
+
+    put = new Put(Bytes.toBytes(2));
+    put.add("film_id", "1");
+    put.add("film_name", "matrix");
+    put.add("category_name", "thriller");
+    filmCategoryTable.put(put);
+
+    put = new Put(Bytes.toBytes(3));
+    put.add("film_id", "2");
+    put.add("film_name", "equilibrium");
+    put.add("category_name", "action");
+    filmCategoryTable.put(put);
+    filmCategoryManager.flush();
+
+    // run the pipeline
+    WorkflowManager workflowManager = appManager.getWorkflowManager(SmartWorkflow.NAME);
+    workflowManager.start();
+    workflowManager.waitForFinish(5, TimeUnit.MINUTES);
+
+    DataSetManager<TimePartitionedFileSet> outputManager = getDataset(joinedDatasetName);
+    TimePartitionedFileSet fileSet = outputManager.get();
+    List<GenericRecord> outputRecords = readOutput(fileSet, outputSchema);
+
+    Assert.assertEquals(5, outputRecords.size());
+
+    Assert.assertEquals("1", outputRecords.get(0).get("film_id").toString());
```

Comment:
Is there anything guaranteeing order of output records?  I don't think there is, which means you should be comparing sets